### ISSUE TITLE
[Matchmaker] Porting and Overhaul

### ISSUE THREAD

**Issue by MrBlizzard** [MrBlizzard/RCAdmins-Tracker#104]
Here are suggested changes, max approved:

The goal for Matchmaker should successfully model itself after [Pokemon showdown](https://play.pokemonshowdown.com/).
 
Unless specifically mentioned, Matchmaker should function the same as it was pre-1.10.2. If more info is needed, I can provide documentation on how it was before.

We plan to have Seasons in Matchmaker, with this upcoming one being Season 2. For some of the things suggested below (making point attribute in InvMenus, and possibly global match leaderboard), we could push them off until Season 3 if they are too difficult to attain.

The formula in which the win/loss points are calculated should remain the same. The same calculation should be in place, but instead of a 10 starting point, it would be whatever specified below (30 points). So, the general goal is to have whatever one person gains, be the same that the other person loses.

### General Changes
- Raise MMR Gained from 10 points normally to 30 points.
- Starting MMR lowered from 1500 to 1000, and you can't go below 1000. So if you are a 1000 MMR player, you won't lose MMR from loss. 
- Set a minimum gain and loss to 5 points. EX: If you're a 2000 MMR Player, and you get matched with a 1000 MMR Player, you won't be able to gain less than 5 points on win (pre-1.10.2, the min gain was 1 point, terribly difficult to raise MMR). If you're the 1000 MMR player in the scenario, you would lose a min of 5 points (otherwise people would be losing 1 point making it unequal).
- Set Max Point loss to 5x whatever you stand to gain. EX: You're able to gain 5 points from your battle upon win. If you lose, the maximum you could lose is 25 points (in the last version, you were standing to gain 1 point, and lose 20, making it not worth battling at higher MMRs).
- Take into account whether the player left manually or due to a crash. If crash, the battle should end in a stalemate. If manually logged out, then the battle should end in a loss for the player who logged off.
- Add a cooldown timer message. Players will know exactly how long they have to wait until they can join match again.
- Add a global match leaderboard, taking people's MMR from all pixels and making a global leaderboard with everyone across pixels on it (Not needed if too hard to achieve, is more cosmetic then influential).
- Phase out Pointshop and Rankshop configs completely and implement a point attribute into InventoryMenus. Any menu imaginable could have the attribute `MMR`, or with the separation in mind `WildMMR:#`, with the minimum being set there. (Mainly for ease of configuration and compatibility, and could fix small issues. This is low priority, and if it's not easily attainable, it's fine to be left out until a later date)
- Take into account #33, where the battle sometimes randomly ended in a stalemate, and try to fix it.

### Two Different Ladders
TM0's are a gift added specifically on RC that allows people to put crazy moves on certain Pokemon unnaturally. This drastically taints the normal competitive atmosphere. The proposed way to keep the competitive nature is making two ladders. One for a "Wild" ladder, where these TM0'd Pokemon and legends could be used, and then a "Normal" Ladder. (Both names could be changed to cooler stuff, just place holders).

This would require:

- The "/match join" command be modified to require the player to specify queue they want to join (ex: /match join wild)

- The "/leaderboard" command be modified to include two separate leaderboards, one for wild and one for normal

- Two separate cooldowns, one for wild and one for normal

- Two separate MMR Points to be made. Wild MMR and Normal MMR. This is to separate Pointshop and Rankshop

- Pointshop and Rankshop be edited to take into account both wild and normal MMR

#### Normal Ladder
For the normal, there should be no legends or TM0'd Pokemon able to battle. This is where things are a little tricky.

Ideally, we should have a system to read a Pokemon's moves, and then acknowledge if they are able to be learned by the Pokemon through level up, by TM, or egg moves. Then if a Pokemon in a player's party has a TM0 move (a move not able to be learned by the means specified), they wouldn't be allowed to join.

For legendaries, there's a clause in the game definable in "Battle Rules". You can see the "Battle Rules" menu when trying to start a PvP Pokemon battle with another trainer. The clause is called "Legendary Clause". It should be simple to set the battle rules for these match battles before starting. 

In normal, there are several clauses to be added to these battles to discourage from a non-competitive atmosphere

Clauses to be used in normal: Legendary Clause, Species Clause, Bag Clause, Endless Battle Clause, Sleep Clause, OHKO Clause, Evasion Clause.

With these clauses, the only thing needing to be done manually is the TM0 contingency in the first paragraph.

#### Wild Ladder
Anything goes, No clauses, TM0s are allowed normally, same with legends.
